mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -62,6 +62,43 @@ void ConcatOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<ConcatOptimization>(context);
 }
 
+struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
+  using OpRewritePattern<tosa::PowOp>::OpRewritePattern;
+  // Pattern that matches a Sqrt + Reciprocal to make it into a rsqrt.
+  // Sqrt is represented in tosa by a Pow so we check for Pow + reciprocal.
+  LogicalResult matchAndRewrite(tosa::PowOp op,
+                                PatternRewriter &rewriter) const override {
+    // Check that the PowOp has a single user
+    if (!op->hasOneUse())
+      return failure();
+
+    Operation* user = *op->user_begin();
+    // Check that this user is a reciprocal
+    if (!isa<tosa::ReciprocalOp>(user))
+      return failure();
+
```

Comment:
`tosa.pow` allows broadcasting, e.g., we could have
```
  %0 = "tosa.const"() <{value = dense<5.000000e-01> : tensor<1x1x7x1xf32>}> : () -> tensor<1x1x7x1xf32>
  %1 = "tosa.pow"(%arg0, %0) : (tensor<1x5x1x1xf32>, tensor<1x1x7x1xf32>) -> tensor<1x5x7x1xf32>
  %2 = "tosa.reciprocal"(%1) : (tensor<1x5x7x1xf32>) -> tensor<1x5x7x1xf32>
```
in this case, I think this would generate
```
"tosa.rsqrt"(...) : (tensor<1x5x1x1xf32>) -> tensor<1x5x7x1xf32>
```
which is not valid. You can either extend this pattern to account for broadcasting (I guess it would need to insert a `tile`), or you just check that the input type of the first pow argument is equal to the result type of the pow (seems more reasonable to me).
(location for this comment is not important)